Add plotting script for benchmark

masterleinad opened the pull request:
This is a reduced version of @aprokop's script used for the paper. The intent here is that we can pass any kind of (partial) results produced by `benchmarks/bvh_driver/ArborX_BoundingVolumeHierarchy.exe` and display them graphically. At the moment, we print the mean with some error base indicating a transformed confidence interval for the rate with which points are processed.
Optionally, the actually numbers are also printed. Sample output:
[
![plot_newest_errorbars](https://user-images.githubusercontent.com/2423533/95247289-ae68b380-07e3-11eb-8671-7eb3a3791d5e.png)
](url)

Feel free to comment on how to improve it.

aprokop commented on scripts/plot.py:
```diff
@@ -0,0 +1,147 @@
+
+def libraries_comparison_rate_figure(input_file, algorithm, geometry, numbers):
+    # Phase 0: parse backends
+    backends = set()
+    with open(input_file) as f:
+        for line in f:
+            search_result = re.search("[^<]*<ArborX::BVH<([^>]*)>.*", line)
+            if search_result != None:
+                backends.add(search_result.group(1))
+    backends = list(backends)
+    backends.sort()
```
I like the automatic way. I also think it should have a manual mode. For example, you may want to get a subset of backends to plot.

aprokop commented on scripts/plot.py:
```diff
@@ -0,0 +1,147 @@
+    width = 0.6
+    scale = 1/len(backends)
+    for i in range(0, len(backends)):
+        barplot = ax.bar(np.arange(len(unified_num_sources)) + scale*(i-1.5) * width,
```
This is the tricky part. The magic number 1.5 really depends on the number of backends. You would want to make a group of columns centered on the tick.

aprokop commented on scripts/plot.py:
```diff
@@ -0,0 +1,147 @@
+  -g GEOMETRY --geometry GEOMETRY   Geometry (source cloud) ['filled_box', 'hollow_box']
+  -i FILE --input-files=FILE        Input file containing benchmark results
+  -o FILE --output-file=FILE        Output file [default: plot.png]
+  -p PLOT_INDEX --plot=PLOT_INDEX   Plot index
```
What's plot index? Is it used?

aprokop commented on scripts/plot.py:
```diff
@@ -0,0 +1,147 @@
+        for j in range(0, len(num_sources[i])):
+            index = unified_num_sources.index(num_sources[i][j])
+            data[i, index] = (num_sources[i][j] / timings[i][j])/10**6
+            errors_data[i, 0, index] = num_sources[i][j]/10**6*(1/timings[i][j] - 1/(timings[i][j]+errors[i][j]))
+            errors_data[i, 1, index] = num_sources[i][j]/10**6*(1/(timings[i][j]-errors[i][j]) - 1/timings[i][j])
```
Hmm, this should be fixed.

masterleinad replied:
What else do you want?

aprokop commented on scripts/plot.py:
```diff
@@ -0,0 +1,147 @@
+    ax.set_ylabel('Rate (million objects/second)', fontsize=20)
+    plt.yticks(fontsize=18)
+    ax.yaxis.grid('on')
+    plt.title('Construction', fontsize=18)
```
This should not be hardcoded, and instead should be mapped from the `algorithm`.

aprokop reviewed (changes requested):
Good start. 

Please make me a co-author (at least, on the first commit).

aprokop commented:
I also wanted to ask. I assume that CUDA and AMD will be run on different machines, and thus produce different output files. I did not see anything in the script that allows to combine multiple files. Would that have to be done manually prior to running the script?

masterleinad commented:
> I also wanted to ask. I assume that CUDA and AMD will be run on different machines, and thus produce different output files. I did not see anything in the script that allows to combine multiple files. Would that have to be done manually prior to running the script?

I am not quite sure what all the actual use cases are. So far, I was assuming to just read a single file but it should not be difficult to read multiple if that is more convenient.

aprokop commented on scripts/plot.py:
```diff
@@ -0,0 +1,137 @@
+    for i in range(0, len(backends)):
+        for j in range(0, len(num_sources[i])):
+            index = unified_num_sources.index(num_sources[i][j])
+            data[i, index] = (num_sources[i][j] / timings[i][j])/10**6
+            errors_data[i, 0, index] = num_sources[i][j]/10**6*(1/timings[i][j] - 1/(timings[i][j]+errors[i][j]))
+            errors_data[i, 1, index] = num_sources[i][j]/10**6*(1/(timings[i][j]-errors[i][j]) - 1/timings[i][j])
```
I think Damien is correct. Looking at the [propagation of uncertainty](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae) for `f = A/B`, the deviation of the inverse `1/B` is indeed `dev(B)/B^2`. So, I think it would be sufficient to do this:
```suggestion
			scaling_factor = num_sources[i][j]/10**6
            data[i, index] = scaling_factor / timings[i][j]
            error = errors[i][j]/(timings[i][j]**2)
            errors_data[i, 0, index] = data[i, index] - scaling_factor * error
            errors_data[i, 1, index] = data[i, index] + scaling_factor * error
```
While it is still not exactly correct (as we are using medians), I think it should be closer. 

P.S. I have not tested this, and it may contain misprints.

masterleinad replied:
For me the question is how you want to interpret these bars and I normally think of them as some kind of confidence interval. I could be convinced that the original distribution is normal-like but the inverse is certainly not so I don't know how much value providing a standard deviation has.
In practice, the difference should be small anyway. 
If both of you rather want the formula above, tell me again and I will change.

aprokop commented:
I know it may annoy you, but there is one more thing. For people who have trouble distinguishing colors (I know at least couple), it is generally helpful to provide some markings on the bars in addition to colors.

masterleinad commented:
@masterleinad I think I addressed all your suggestions apart from the error bars now. I am assuming to squash all the commits in the end anyway so only added you as co-author to the last one even if it is not directly related (to avoid force-pushing early).

aprokop commented:
@masterleinad Thanks! I am not going to argue about errors right now. Lets wait to see if @dalg24 has comments, while i take final look at the rest of the pr.

aprokop commented on scripts/plot.py:
```diff
@@ -0,0 +1,151 @@
+    else:
+        backends = backend_list
+
+    # Phase 1: data setup
+    num_sources  = []
+    timings = []
+    errors = []
+    unified_num_sources = [];
+    for i in range(len(backends)):
+        tmp_num_sources = []
+        tmp_timings = []
+        tmp_errors = []
+        for input_file in input_files:
+            tmp_file_num_sources, _, tmp_file_timings = parse_benchmark(input_file, algorithm, 'ArborX::BVH<'+backends[i]+'>', 'median', geometry)
+            tmp_file_num_sources, _, tmp_file_errors = parse_benchmark(input_file, algorithm, 'ArborX::BVH<'+backends[i]+'>', 'stddev', geometry)
+            tmp_num_sources = tmp_num_sources + tmp_file_num_sources
+            tmp_timings = tmp_timings + tmp_file_timings
+            tmp_errors = tmp_errors + tmp_file_errors
+        num_sources.append(tmp_num_sources)
+        timings.append(tmp_timings)
+        errors.append(tmp_errors)
+        unified_num_sources = unified_num_sources + tmp_num_sources
+
+    unified_num_sources = list(set(unified_num_sources))
+    unified_num_sources.sort()
+    num_source_labels = [None] * len(unified_num_sources)
+    for i in range(len(unified_num_sources)):
+        num_source_labels[i] = '$10^' + str(round(math.log(unified_num_sources[i], 10))) + '$'
```
I'm struggling to understand what this does. Could you elaborate on what the end result of `num_sources`, `timings` or `errors` is going to be if there are repetitions in multiple files?

masterleinad replied:
If there are repetitions only the last run for a specific setting is plotted. This is actually not related to the code above but rather to how we initialize `data` below.

aprokop commented on scripts/plot.py:
```diff
@@ -0,0 +1,151 @@
+    if algorithm != 'construction' and algorithm != 'knn_search' and algorithm != 'radius_search':
+        raise Exception('Unknown algorithm: "' + algorithm + '"')
+
+    backends = backends.split(',')
+    input_files = input_files[0]
+    input_files = input_files.split(',')
```
I think that if you use docopt's `INPUT...`, you can do `-i file1 -i file2 -i files3` automatically producing a list, so you don't need this.

masterleinad replied:
I can try that. Specifying input files via `-i file1,file2,file3` was my first intuition so I tried to make that work. I don't care much one way or the other, though.

masterleinad replied:
Yes, commenting these two lines `-i file1 -i file2 -i files3` works. Do you prefer that syntax?

aprokop replied:
Yes, this way we rely on docopt and no write the code ourselves. Could do the same with backends.

masterleinad replied:
Why would you design interfaces rather for ease of implementation than for ease of use?
Anyway, I don't feel strongly it so I'll just change.

masterleinad replied:
Here you go!

aprokop commented:
I think what needs to be done here to finally merge it is:
- Switch to parsing JSON files instead of regular output
  It should be more robust, and would simplify the code through removing `benchmark.py`, relying instead on the regular JSON parsing libraries
- Switch to using rates introduced in #457.

@dalg24 You had a piece of code somewhere already to parse benchmark json files already, correct? Can you please share?

aprokop commented:
![Screen Shot 2021-01-11 at 9 34 36 AM](https://user-images.githubusercontent.com/7297887/104195133-3ae0ce80-53f0-11eb-8a31-dbc7b38571ef.png)

Woah, did not realize that one can also close pull requests automatically from another pull request. Thought it only affected issues.

aprokop commented:
I updated the benchmark plotting to use rates introduced in #457 and parse JSON files instead of text output. I think this is ready for final review.

aprokop commented:
Aggregate names and runtimes were [added](https://github.com/google/benchmark/commit/58588476ce52c0af4cb54f9c596c4579f78bc953) in v1.5.

aprokop commented:
One thing that could be nice, but is entirely optional, is to try to parse file exactly once. I'm thinking we could call `populate_data()` and produce lists (including the backends) first, and then trim the lists depending on the `backends` argument. When plotting results from larger runs with multiple file, I noticed that it may takes 10+ seconds for a plot, and couple minutes to cycles through all combinations of algorithms and geometries. Would be nice to bring it down to somewhat interactive levels.

masterleinad commented on scripts/benchmark_plot.py:
```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env python3
+"""plot.py
+
+Usage:
+  plot.py -a ALGORITHM [-b BACKENDS...] -g GEOMETRY -i INPUT... [-n] [-e] [-o OUTPUT_FILE]
+  plot.py (-h | --help)
```
```suggestion
"""benchmark_plot.py

Usage:
  benchmark_plot.py -a ALGORITHM [-b BACKENDS...] -g GEOMETRY -i INPUT... [-n] [-e] [-o OUTPUT_FILE]
  benchmark_plot.py (-h | --help)
```

masterleinad commented on scripts/benchmark.py:
```diff
@@ -0,0 +1,67 @@
+import json
+import re
+
+def load_json(filename):
+    with open(filename) as f:
+        data = json.loads(f.read())
+    return data
```
What is the advantage of requiring `json`? IMHO it's easier to just use the plain format since this is created by the benchmark by default.

masterleinad replied:
I checked that it works but I don't necessarily want to create the output on hte same nodes that I am running and copying the much shorter plain format it just easier.

aprokop replied:
For me, the advantage of using JSON is that we already have to dump it if we want to use `compare_bench.py` script for run comparison. But you bring a valid point that removing the old "one-line" console output may be stupid. I feel that it brings few points worth discussing:
- With JSON, I believe that we could figure out a way to parse each file exactly once, as we are able to store file as a dictionary in a variable. I think this is challenging with console output
- The rates output is done in human form, using `k` or `M` suffix to indicate thousands/millions. This leads to extra complexity.

This is not to say that I'm opposed to bringing parsing output back. It does have its benefit in its simplicity.

aprokop replied:
@masterleinad I added the console output file reader.

masterleinad commented on scripts/benchmark_plot.py:
```diff
@@ -0,0 +1,179 @@
+
+    # Check input
+    if geometry not in ['filled_box', 'hollow_box']:
+        raise Exception('Unknown geometry: "' + geometry + '"')
```
```suggestion
        raise Exception('Unknown geometry: "' + geometry + '". Allowed values are "filled_box" and "hollow_box".')
```

aprokop replied:
Allowed values are seen in `./benchmark_plot.py --help`, no need for them here.

masterleinad replied:
I know. Just my user experience: If I misspell an option so I prefer knowing what the correct spelling/allowed options are immediately without needing to relaunch with `--help`.

masterleinad commented on scripts/benchmark_plot.py:
```diff
@@ -0,0 +1,179 @@
+        raise Exception('Unknown geometry: "' + geometry + '"')
+
+    if algorithm not in ['construction', 'knn_search', 'radius_search']:
+        raise Exception('Unknown algorithm: "' + algorithm + '"')
```
```suggestion
        raise Exception('Unknown algorithm: "' + algorithm + '". Allowed values are "construction", "knn_search" and "radius_search"')
```

aprokop replied:
Allowed values are listed with `--help`, no need for them here.

masterleinad replied:
Are you fine with me adding this myself in a follow-up pull request?

aprokop replied:
Oh, I forgot about these. Just push to this PR.

aprokop commented:
I think the best way to address reading regular text output is to provide a utility function in plotting that is able to convert it to json. This way, there would be no code overlap with JSON reader, and it would be agnostic to the actual data.

aprokop commented:
OK, this does what I want it to. Still has some rough edges, and certainly places could use some love, but lets try to work on merging it. In the current form, it is extremely useful, and I've been playing with it all day producing figures.

dalg24 commented on scripts/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+docopt
+matplotlib
+numpy
+json
```
```suggestion
```

aprokop replied:
why?

dalg24 replied:
because it is a bug.  it is part of the python standard library.  if you try to install the requirements you will get an error

aprokop replied:
OK, understood now. `requirements.txt` only to contain non-core libraries. Surprisingly to me, `json` is part of core.

masterleinad reviewed (commented):
I'm fine with moving forward here.

aprokop commented:
@masterleinad You missed couple new options that I added like `radius_callback_search` and `knn_callback_search`. That's why I said we should have a single place with all the options somewhere, so that we won't run into this.

dalg24 commented on scripts/benchmark.py:
```diff
@@ -36,11 +36,16 @@ def load_json(filename):
 
     return data
 
+allowed_geometries = ['filled_box', 'hollow_box']
```
`geometries.keys()`

aprokop replied:
Don't understand.

aprokop replied:
OK, understood now. `requirements.txt` only to contain non-core libraries. Surprisingly to me, `json` is part of core.

aprokop commented:
@dalg24 Did you have other comments?

aprokop commented:
I terminated the testing (forgot to add `[ci skip]` to the commit message).
